<?xml version='1.0'?>
<robot name="proj1robot">
  <!-- Give the robot a color -->
  <material name="robotcolor">
    <color rgba="1 0.757 0.147 1"/>
  </material>

  <!-- root link, on the ground just below the model origin -->
  <link name="base_footprint">
   <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.001 0.001 0.001" />
      </geometry>
    </visual>
  </link>

  <joint name="base_link_joint" type="fixed">
    <origin xyz="0.0 0 0" rpy="0 0 0" />
    <parent link="base_footprint"/>
    <child link="base_link" />
  </joint>

  <!-- the model -->
  <link name="base_link">
    <inertial>
      <mass value="50" />
      <origin xyz="0 0.2 0" />
      <inertia ixx="50.0" ixy="0.0" ixz="0.0"
        iyy="50.0" iyz="0.0"
        izz="50.0" />
    </inertial>
    <visual>
      <origin xyz="0 0.2 0.1" />
      <geometry>
        <box size="0.2 0.4 0.2" />
      </geometry>
      <material name="robotcolor"/>
    </visual>
    <collision>
      <origin xyz="0 0.2 0.1" />
      <geometry>
        <box size="0.2 0.4 0.2" />
      </geometry>
    </collision>
  </link>

  <!-- planar move, from gazebo plugins page -->
  <gazebo>
    <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
      <commandTopic>robot/cmd_vel</commandTopic>
      <odometryTopic>robot/move_odom</odometryTopic>
      <odometryFrame>proj1robot_odom</odometryFrame>
      <odometryRate>20.0</odometryRate>
      <robotBaseFrame>base_footprint</robotBaseFrame>
    </plugin>
  </gazebo>

  <!-- colorize the robot -->
  <gazebo reference="base_link">
    <material>Gazebo/DarkYellow</material>
  </gazebo>

  <!-- Get a ground truth publisher -->
  <gazebo reference="base_link">
    <plugin name="p3d_base_controller" filename="libgazebo_ros_p3d.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>10.0</updateRate>
      <bodyName>base_link</bodyName>
      <topicName>robot/ground_truth</topicName>
      <gaussianNoise>0.0</gaussianNoise>
      <frameName>world</frameName>
      <xyzOffsets>0 0 0</xyzOffsets>
      <rpyOffsets>0 0 0</rpyOffsets>
    </plugin>
  </gazebo>

  <!-- get an imu publisher -->
  <gazebo reference="base_link">
    <gravity>true</gravity>
    <sensor name="imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>imu</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>robot/imu</topicName>
        <bodyName>base_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0.2 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>proj1robot_imu</frameName>
        <initialOrientationAsReference>false</initialOrientationAsReference>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
  </gazebo>
</robot>
